app/assets/javascripts/discourse/app/lib/codeblock-buttons.js

```diff
@@ -0,0 +1,193 @@
+
+    this._createButtons(codeBlocks);
+    const commandId = guidFor(element);
+    this._storeClickHandler(commandId, element);
+    this._addClickEvent(element);
+  }
+
+  cleanup() {
+    Object.values(this._codeblockButtonClickHandlers || {}).forEach((handler) =>
+      handler.removeEventListener("click", this._boundClickHandler)
+    );
+
+    Object.values(this._fadeCopyCodeblocksRunners || {}).forEach((runner) =>
+      cancel(runner)
+    );
+
+    this._codeblockButtonClickHandlers = {};
+    this._fadeCopyCodeblocksRunners = {};
+  }
+
+  _storeClickHandler(identifier, element) {
+    if (this._codeblockButtonClickHandlers[identifier]) {
+      this._codeblockButtonClickHandlers[identifier].removeEventListener(
+        "click",
+        this._boundClickHandler
+      );
+
+      delete this._codeblockButtonClickHandlers[identifier];
+    }
+
+    this._codeblockButtonClickHandlers[identifier] = element;
+  }
+
+  _getCodeBlocks(element) {
+    let codeBlocks = [];
+    try {
+      codeBlocks = element.querySelectorAll(
+        ":scope > pre > code, :scope :not(article):not(blockquote) > pre > code"
+      );
+    } catch (e) {
+      // :scope is probably not supported by this browser
+      codeBlocks = [];
+    }
+    return codeBlocks;
+  }
+
+  _createButtons(codeBlocks) {
+    codeBlocks.forEach((codeBlock) => {
+      if (this.showFullscreen) {
+        const fullscreenButton = document.createElement("button");
+        fullscreenButton.classList.add("btn", "nohighlight", "fullscreen-cmd");
+        fullscreenButton.innerHTML = iconHTML("discourse-expand");
+        codeBlock.before(fullscreenButton);
+      }
+
+      if (this.showCopy) {
+        const copyButton = document.createElement("button");
+        copyButton.classList.add("btn", "nohighlight", "copy-cmd");
+        copyButton.innerHTML = iconHTML("copy");
+        codeBlock.before(copyButton);
+      }
+
+      codeBlock.parentElement.classList.add("codeblock-buttons");
+    });
+  }
+
+  _addClickEvent(element) {
+    element.addEventListener("click", this._boundClickHandler, false);
+  }
+
+  _handleClick(event) {
+    if (
+      !event.target.classList.contains("copy-cmd") &&
+      !event.target.classList.contains("fullscreen-cmd")
+    ) {
+      return;
+    }
+
+    const action = event.target.classList.contains("fullscreen-cmd")
+      ? "fullscreen"
+      : "copy";
+    const button = event.target;
+    const codeEl = button.parentElement.querySelector("code");
+
+    if (codeEl) {
+      // replace any weird whitespace characters with a proper '\u20' whitespace
+      const text = codeEl.innerText
+        .replace(
+          /[\f\v\u00a0\u1680\u2000-\u200a\u202f\u205f\u3000\ufeff]/g,
+          " "
+        )
+        .trim();
+
+      if (action === "copy") {
+        const result = clipboardCopy(text);
+        if (result.then) {
+          result.then(() => {
+            this._copyComplete(button);
+          });
+        } else if (result) {
```

Comment:
The order here is a bit weird, we check if result is a promise and _then_ we check if there is a result. Shouldn't we check first that there is a result?

Reply:
The first `if` should read:

```
if (result?.then) {
```

Reply:
This was in the original copy-codeblocks code, will change